File: ruby/benchmark.rb
# frozen_string_literal: true

require 'benchmark'
require_relative 'lua_deflate'
require_relative 'lua_deflate_native'

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def random_printable(size)
  Array.new(size) { rand(32..126).chr }.join
end

SIZES = {
  '32 B'   =>   32,
  '1 KB'   => 1_024,
  '64 KB'  => 65_536,
  '256 KB' => 262_144
}.freeze

# ---------------------------------------------------------------------------
# Correctness
# ---------------------------------------------------------------------------

puts "=== Correctness ==="
puts

SIZES.each do |label, size|
  data = random_printable(size)

  enc_ref    = LuaDeflate.encode_for_print(data)
  enc_native = LuaDeflateNative.encode_for_print(data)

  enc_match = enc_ref == enc_native

  dec_ref_from_ref       = LuaDeflate.decode_for_print(enc_ref)
  dec_native_from_native = LuaDeflateNative.decode_for_print(enc_native)
  dec_ref_from_native    = LuaDeflate.decode_for_print(enc_native)
  dec_native_from_ref    = LuaDeflateNative.decode_for_print(enc_ref)

  roundtrip_ref    = dec_ref_from_ref    == data
  roundtrip_native = dec_native_from_native == data
  cross_ref        = dec_ref_from_native == data
  cross_native     = dec_native_from_ref == data

  all_pass = enc_match && roundtrip_ref && roundtrip_native && cross_ref && cross_native
  status   = all_pass ? 'PASS' : 'FAIL'

  puts "#{label.rjust(7)}: #{status}"
  unless all_pass
    puts "  enc match:      #{enc_match}"
    puts "  roundtrip ref:  #{roundtrip_ref}"
    puts "  roundtrip nat:  #{roundtrip_native}"
    puts "  cross ref:      #{cross_ref}"
    puts "  cross native:   #{cross_native}"
  end
end

puts

# ---------------------------------------------------------------------------
# Benchmark
# ---------------------------------------------------------------------------

# Pick iteration counts so each cell takes ~0.5–2 s on typical hardware.
ITER_MAP = {
  '32 B'   => 20_000,
  '1 KB'   =>  2_000,
  '64 KB'  =>    50,
  '256 KB' =>    12
}.freeze

puts "=== Benchmark (encode + decode round-trip) ==="
puts

col_w = 8
printf "%-9s %#{col_w}s %#{col_w}s %#{col_w}s %#{col_w}s %#{col_w}s\n",
       'Size', 'Iters', 'Ref(ops/s)', 'Nat(ops/s)', 'Speedup', ''
puts '-' * 60

SIZES.each do |label, size|
  data   = random_printable(size)
  iters  = ITER_MAP[label]

  # Warm up
  3.times do
    LuaDeflate.encode_for_print(data)
    LuaDeflateNative.encode_for_print(data)
  end

  t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  iters.times do
    enc = LuaDeflate.encode_for_print(data)
    LuaDeflate.decode_for_print(enc)
  end
  ref_elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0

  t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  iters.times do
    enc = LuaDeflateNative.encode_for_print(data)
    LuaDeflateNative.decode_for_print(enc)
  end
  nat_elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0

  ref_ops = iters / ref_elapsed
  nat_ops = iters / nat_elapsed
  speedup = nat_ops / ref_ops

  flag = speedup >= 1.0 ? 'faster' : 'slower'
  printf "%-9s %#{col_w}d %#{col_w}.0f %#{col_w}.0f %#{col_w}.2fx %s\n",
         label, iters, ref_ops, nat_ops, speedup, flag
end

puts

File: js/benchmark.js
'use strict';

const LuaDeflate = require('./lib/LuaDeflate');
const LuaDeflateNative = require('./lib/LuaDeflateNative');

const orig = new LuaDeflate();
const native = new LuaDeflateNative();

// Generate random printable ASCII string (0x20-0x7E)
function randomString(len) {
    let s = '';
    for (let i = 0; i < len; i++) {
        s += String.fromCharCode(0x20 + Math.floor(Math.random() * 95));
    }
    return s;
}

const SIZES = [
    { label: 'Small  (32B)', bytes: 32,     encIter: 5000, decIter: 5000 },
    { label: 'Medium (1KB)', bytes: 1024,   encIter: 1000, decIter: 1000 },
    { label: 'Large  (64KB)', bytes: 65536,  encIter: 20,   decIter: 20   },
    { label: 'XL   (256KB)', bytes: 262144, encIter: 5,    decIter: 5    },
];

// --- Correctness check ---
console.log('=== Correctness ===\n');

let allPass = true;
for (const { label, bytes } of SIZES) {
    const input = randomString(bytes);

    const encOrig   = orig.encodeForPrint(input);
    const encNative = native.encodeForPrint(input);

    const encMatch = encOrig === encNative;

    const decOrig   = orig.decodeForPrint(encOrig);
    const decNative = native.decodeForPrint(encNative);

    const decOrigMatch   = decOrig   === input;
    const decNativeMatch = decNative === input;

    const pass = encMatch && decOrigMatch && decNativeMatch;
    if (!pass) allPass = false;

    const status = pass ? 'PASS' : 'FAIL';
    console.log(`${label}: encode-match=${encMatch} dec-orig=${decOrigMatch} dec-native=${decNativeMatch} => ${status}`);
    if (!encMatch) {
        console.log(`  enc orig   len=${encOrig   ? encOrig.length   : 'null'}`);
        console.log(`  enc native len=${encNative ? encNative.length : 'null'}`);
    }
}
console.log(`\nOverall correctness: ${allPass ? 'PASS' : 'FAIL'}\n`);

// --- Benchmark helper ---
function bench(fn, iterations) {
    // Warm up
    fn();
    fn();

    const start = process.hrtime.bigint();
    for (let i = 0; i < iterations; i++) {
        fn();
    }
    const end = process.hrtime.bigint();
    const elapsedNs = Number(end - start);
    const opsPerSec = Math.round(iterations / (elapsedNs / 1e9));
    return opsPerSec;
}

// --- Benchmark ---
console.log('=== Benchmark ===\n');

const COL = {
    label:   22,
    op:      8,
    orig:    14,
    native:  14,
    speedup: 10,
};

function pad(str, width) {
    return String(str).padStart(width);
}

function fmt(n) {
    return n.toLocaleString('en-US');
}

const header = [
    'Size'.padEnd(COL.label),
    'Op'.padEnd(COL.op),
    pad('Original', COL.orig),
    pad('Native', COL.native),
    pad('Speedup', COL.speedup),
].join('  ');

const divider = '-'.repeat(header.length);

console.log(header);
console.log(divider);

for (const { label, bytes, encIter, decIter } of SIZES) {
    const input    = randomString(bytes);
    const encoded  = orig.encodeForPrint(input);

    const encOrig   = bench(() => orig.encodeForPrint(input),     encIter);
    const encNative = bench(() => native.encodeForPrint(input),   encIter);
    const decOrig   = bench(() => orig.decodeForPrint(encoded),   decIter);
    const decNative = bench(() => native.decodeForPrint(encoded), decIter);

    const encSpeedup = (encNative / encOrig).toFixed(2);
    const decSpeedup = (decNative / decOrig).toFixed(2);

    console.log([
        label.padEnd(COL.label),
        'encode'.padEnd(COL.op),
        pad(fmt(encOrig),   COL.orig),
        pad(fmt(encNative), COL.native),
        pad(`${encSpeedup}x`, COL.speedup),
    ].join('  '));

    console.log([
        ''.padEnd(COL.label),
        'decode'.padEnd(COL.op),
        pad(fmt(decOrig),   COL.orig),
        pad(fmt(decNative), COL.native),
        pad(`${decSpeedup}x`, COL.speedup),
    ].join('  '));

    console.log(divider);
}

console.log('\nops/sec: higher is better. Speedup = native / original.\n');
